File: src/main/java/de/hannisoft/gaveplan/excelimport/GraveFileReader.java
package de.hannisoft.graveplan.excelimport;

import java.io.File;
import java.io.IOException;
import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Map;
import java.util.TreeMap;

import de.hannisoft.graveplan.model.Grave;
import de.hannisoft.graveplan.model.GraveMap;
import de.hannisoft.graveplan.model.GraveSite;
import jxl.Sheet;
import jxl.Workbook;
import jxl.WorkbookSettings;
import jxl.read.biff.BiffException;

public class GraveFileReader extends AbstractXLSFileReader {
    private static final String COL_GARVE_ID = "Grabstätte";
    private static final String COL_PLACE = "Grabstelle";
    private static final String COL_LAST_NAME = "Nachname";
    private static final String COL_FIRST_NAME = "Vorname";
    private static final String COL_DOD = "Sterbedatum";
    private static final String COL_DOB = "Geburtsdatum";
    private static final String COL_FUNERAL_DATE = "Bestattungsdatum";

    public Map<String, GraveMap> read(String inputFile, final Map<String, GraveSite> graveSites) throws IOException {
        Map<String, GraveSite> emptyGraveSites = new TreeMap<String, GraveSite>(graveSites);
        Map<String, GraveMap> graveMaps = new TreeMap<>();
        File inputWorkbook = new File(inputFile);
        Workbook w;
        WorkbookSettings ws = new WorkbookSettings();
        ws.setEncoding("Cp1252");
        try {
            w = Workbook.getWorkbook(inputWorkbook, ws);
            Sheet sheet = w.getSheet(0);
            initColumnMap(sheet, 1);
            DateFormat format = new SimpleDateFormat("dd.MM.yyyy");
            for (int i = 2; i < sheet.getRows(); i++) {
                String graveId = getContent(COL_GARVE_ID, i);
                emptyGraveSites.remove(graveId);
                GraveSite graveSite = graveSites.get(graveId);
                if (graveSite != null) {
                    String[] graveString = getContent(COL_PLACE, i).split("\\/");
                    String row = graveString[0];
                    String placeNo = graveString[1];
                    String lastName = getContent(COL_LAST_NAME, i);
                    String firstName = getContent(COL_FIRST_NAME, i);
                    String dateOfDeath = getContent(COL_DOD, i);
                    if (dateOfDeath == null || dateOfDeath.trim().isEmpty()) {
                        dateOfDeath = getContent(COL_FUNERAL_DATE, i);
                    }
                    String dateOfBirth = getContent(COL_DOB, i);
                    try {
                        Grave grave = new Grave(graveSite, row, placeNo);
                        grave.setDeceased(firstName + " " + lastName);
                        if (dateOfDeath != null && !dateOfDeath.trim().isEmpty()) {
                            grave.setDateOfDeatch(format.parse(dateOfDeath.replace('/', '.')));
                        }
                        if (dateOfBirth != null && !dateOfBirth.trim().isEmpty()) {
                            grave.setDateOfBirth(format.parse(dateOfBirth.replace('/', '.')));
                        }
                        graveSite.getGraves().add(grave);

                        String field = graveSite.getField();
                        GraveMap placeMap = graveMaps.get(field);
                        if (placeMap == null) {
                            placeMap = new GraveMap(field);
                            graveMaps.put(field, placeMap);
                        }
                        placeMap.addGraveSite(graveSite);
                    } catch (ParseException e) {
                        e.printStackTrace();
                    }

                }
            }
        } catch (BiffException e) {
            e.printStackTrace();
        }
        for (GraveSite graveSite : emptyGraveSites.values()) {
            String field = graveSite.getField();
            GraveMap placeMap = graveMaps.get(field);
            if (placeMap == null) {
                placeMap = new GraveMap(field);
                graveMaps.put(field, placeMap);
            }
            placeMap.addGraveSite(graveSite);
        }
        return graveMaps;
    }

}

File: src/main/kotlin/de/hannisoft/graveplan/excelimport/GraveFileReader.kt
package de.hannisoft.graveplan.excelimport

import de.hannisoft.de.hannisoft.graveplan.excelimport.buildHeaderMap
import de.hannisoft.de.hannisoft.graveplan.excelimport.getValue
import de.hannisoft.de.hannisoft.graveplan.excelimport.parseDateString
import de.hannisoft.graveplan.model.Grave
import de.hannisoft.graveplan.model.GraveMap
import de.hannisoft.graveplan.model.GraveSite
import org.apache.poi.ss.usermodel.WorkbookFactory
import java.io.File

class GraveFileReader {
    companion object {
        const val COL_GARVE_ID: String = "Grabstätte"
        const val COL_PLACE: String = "Grabstelle"
        const val COL_LAST_NAME: String = "Nachname"
        const val COL_FIRST_NAME: String = "Vorname"
        const val COL_DOD: String = "Sterbedatum"
        const val COL_DOB: String = "Geburtsdatum"
        const val COL_FUNERAL_DATE: String = "Bestattungsdatum"
    }

    fun read(inputFile: File, graves: Map<String, GraveSite>) {
        val emptyGraveSites = graves.toMutableMap()
        val gravesMaps = mutableMapOf<String, GraveMap>()

        inputFile.inputStream().use { fis ->
            val workbook = WorkbookFactory.create(fis)
            val sheet = workbook.getSheetAt(0)
            var headerMap = sheet.buildHeaderMap()
            var startRow = 1
            if (headerMap.size < 5) {
                // if the header map contains less than 5 columns, there is a "GroupingRow" above the current column header row, which must be ignored
                headerMap = sheet.buildHeaderMap(1)
                startRow = 2
            }
            for (rowIndex in startRow..sheet.lastRowNum) {
                val row = sheet.getRow(rowIndex) ?: continue
                val graveId = row.getValue(headerMap[COL_GARVE_ID])
                try {
                    emptyGraveSites.remove(graveId)
                    graves[graveId]?.apply {
                        val graveString = row.getValue(headerMap[COL_PLACE]).split("/")
                        val rowStr = graveString[0].trim()
                        val placeStr = graveString[1].trim().substring(0, 2)
                        val lastName = row.getValue(headerMap[COL_LAST_NAME])
                        val firstName = row.getValue(headerMap[COL_FIRST_NAME])
                        var dateOfDeath = row.getValue(headerMap[COL_DOD])
                        if (dateOfDeath.isEmpty()) {
                            dateOfDeath = row.getValue(headerMap[COL_FUNERAL_DATE])
                        }
                        val dateOfBirth = row.getValue(headerMap[COL_DOB])
                        val grave = Grave(this, rowStr.toInt(), placeStr.toInt())
                        grave.deceased = "$firstName $lastName"
                        grave.dateOfDeath = parseDateString(dateOfDeath)
                        grave.dateOfBirth = parseDateString(dateOfBirth)
                        this.graves.add(grave)

                        val field = this.field
                        val placeMap = gravesMaps[field] ?: GraveMap(field).also {
                            gravesMaps[field] = it
                        }
                        placeMap.addGraveSite(this)
                    }
                } catch (e: Exception) {
                    System.err.println("${e.javaClass.simpleName} while reading grave $graveId from $inputFile at row $rowIndex: ${e.message}")
                    throw e
                }
            }
        }

        emptyGraveSites.values.forEach {
            val field = it.field
            val placeMap = gravesMaps[field] ?: GraveMap(field).also {
                gravesMaps[field] = it
            }
            placeMap.addGraveSite(it)
        }
    }
}